src/main/java/org/jenkinsci/plugins/durabletask/FileMonitoringTask.java

```diff
@@ -367,7 +398,113 @@ public FilePath getOutputFile(FilePath workspace) throws IOException, Interrupte
+        }
+        return watchService;
+    }
+
+    private static class StartWatching extends MasterToSlaveFileCallable<Void> {
+
         private static final long serialVersionUID = 1L;
+
+        private final FileMonitoringController controller;
+        private final Handler handler;
+        private final TaskListener listener;
+
+        StartWatching(FileMonitoringController controller, Handler handler, TaskListener listener) {
+            this.controller = controller;
+            this.handler = handler;
+            this.listener = listener;
+        }
+
+        @Override public Void invoke(File workspace, VirtualChannel channel) throws IOException, InterruptedException {
+            watchService().submit(new Watcher(controller, new FilePath(workspace), handler, listener));
+            return null;
+        }
+
+    }
+
+    private static class Watcher implements Runnable {
+
+        private final FileMonitoringController controller;
+        private final FilePath workspace;
+        private final Handler handler;
+        private final TaskListener listener;
+        private final @CheckForNull Charset cs;
+
+        Watcher(FileMonitoringController controller, FilePath workspace, Handler handler, TaskListener listener) {
+            this.controller = controller;
+            this.workspace = workspace;
+            this.handler = handler;
+            this.listener = listener;
+            cs = FileMonitoringController.transcodingCharset(controller.charset);
+        }
+
+        @Override public void run() {
+            try {
+                Integer exitStatus = controller.exitStatus(workspace, listener); // check before collecting output, in case the process is just now finishing
+                long lastLocation = 0;
+                FilePath lastLocationFile = controller.getLastLocationFile(workspace);
+                if (lastLocationFile.exists()) {
+                    lastLocation = Long.parseLong(lastLocationFile.readToString());
+                }
+                FilePath logFile = controller.getLogFile(workspace);
+                long len = logFile.length();
+                if (len > lastLocation) {
+                    assert !logFile.isRemote();
+                    try (FileChannel ch = FileChannel.open(Paths.get(logFile.getRemote()), StandardOpenOption.READ)) {
+                        InputStream locallyEncodedStream = Channels.newInputStream(ch.position(lastLocation));
+                        InputStream utf8EncodedStream = cs == null ? locallyEncodedStream : new ReaderInputStream(new InputStreamReader(locallyEncodedStream, cs), StandardCharsets.UTF_8);
+                        try {
+                            handler.output(utf8EncodedStream);
+                        } finally {
+                            lastLocationFile.write(Long.toString(ch.position()), null);
```

Comment:
🐜 Noted that without atomic writes here, if the agent is restarted during writing, then we lose the position within the log and re-send *I think* the entire log.  Potential for looping if somehow we get repeated failures (very unlikely though).

Reply:
What if the handler throws an exception (likely IOException) while writing the output and the output isn't actually saved?  Suppose, for example that we can't connect to whatever is supposed to receive the log output.

Reply:
Wait, at that point will we bomb out completely and never try to send the log again?  Gah, this code is _literally_ giving me headaches to follow the flow of execution because it's so indirect.

@jglick C'mon man, a _few_ more explanatory comments please?

Reply:
Let me say what I *expect*: if the handler throws an exception (probably `IOExeption`) while trying to handle output, or takes longer than some timeout period, we break out (without mutating the lastLocationFile), and retry at least once after an appropriate delay.  

At some point if issues are resolved so the handler passes, we *should* (via mechanism unspecified) be able to get the entire log re-sent from the point of failure, including any new content that was added.

Reply:
> if the handler throws an exception (probably `IOExeption`)

I need to note in a comment that the default handler does not currently ever throw `IOException` as a result of a sink issue (only as a result of a local file error).

Reply:
Changed as of https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/63/commits/62aeb458e7b1da123fc77228c9821ca95a00ddbd.